// @remote-ui/web-workers/build/esm/messenger/iframe/create-messenger.mjs
import { defineProperty as _defineProperty } from '../../_virtual/_rollupPluginBabelHelpers.js';
import { IFRAME_RUN_IDENTIFIER } from './constants.mjs';

function createIframeWorkerMessenger(url, prepareIframe) {
  var _MessageChannel = new MessageChannel(),
      port1 = _MessageChannel.port1,
      port2 = _MessageChannel.port2;

  var iframe = document.createElement('iframe');
  iframe.setAttribute('style', 'display:none;');

  function loadHandler() {
    port1.start();
    iframe.contentWindow.postMessage(_defineProperty({}, IFRAME_RUN_IDENTIFIER, url.href), '*', [port2]);
    iframe.removeEventListener('load', loadHandler);
  }

  iframe.addEventListener('load', loadHandler);
  prepareIframe(iframe);
  document.body.appendChild(iframe);
  return {
    postMessage: function postMessage() {
      return port1.postMessage.apply(port1, arguments);
    },
    addEventListener: function addEventListener() {
      return port1.addEventListener.apply(port1, arguments);
    },
    removeEventListener: function removeEventListener() {
      return port1.removeEventListener.apply(port1, arguments);
    },
    terminate: function terminate() {
      port1.close();
      iframe.remove();
    }
  };
}

export { createIframeWorkerMessenger };
