// @remote-ui/web-workers/build/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var rpc = require('@remote-ui/rpc');
var worker = require('./create/worker.js');
var plainWorker = require('./create/plain-worker.js');
var worker$1 = require('./messenger/worker.js');
var nullOrigin = require('./messenger/iframe/null-origin.js');
var remoteOrigin = require('./messenger/iframe/remote-origin.js');



Object.defineProperty(exports, 'release', {
  enumerable: true,
  get: function () {
    return rpc.release;
  }
});
Object.defineProperty(exports, 'retain', {
  enumerable: true,
  get: function () {
    return rpc.retain;
  }
});
exports.createWorkerFactory = worker.createWorkerFactory;
exports.expose = worker.expose;
exports.terminate = worker.terminate;
exports.createPlainWorkerFactory = plainWorker.createPlainWorkerFactory;
exports.createWorkerMessenger = worker$1.createWorkerMessenger;
exports.createIframeWorkerMessenger = nullOrigin.createNullOriginIframeMessenger;
exports.createNullOriginIframeMessenger = nullOrigin.createNullOriginIframeMessenger;
exports.createRemoteOriginIframeMessengerFactory = remoteOrigin.createRemoteOriginIframeMessengerFactory;
