// @remote-ui/react/build/cjs/host/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@remote-ui/core');
var RemoteRenderer = require('./RemoteRenderer.js');
var RemoteComponent = require('./RemoteComponent.js');
var RemoteText = require('./RemoteText.js');
var controller = require('./controller.js');
var context = require('./context.js');
var hooks = require('./hooks.js');
var workers = require('./workers.js');



Object.defineProperty(exports, 'createRemoteReceiver', {
  enumerable: true,
  get: function () {
    return core.createRemoteReceiver;
  }
});
exports.RemoteRenderer = RemoteRenderer.RemoteRenderer;
exports.RemoteComponent = RemoteComponent.RemoteComponent;
exports.RemoteText = RemoteText.RemoteText;
exports.createController = controller.createController;
exports.ControllerContext = context.ControllerContext;
exports.RemoteReceiverContext = context.RemoteReceiverContext;
exports.useAttached = hooks.useAttached;
exports.useRemoteReceiver = hooks.useRemoteReceiver;
exports.useWorker = workers.useWorker;
