// @remote-ui/react/build/cjs/host/RemoteRenderer.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var jsxRuntime = require('react/jsx-runtime');
var react = require('react');
var core = require('@remote-ui/core');
var hooks = require('./hooks.js');

var RemoteRenderer = /*#__PURE__*/react.memo(function (_ref) {
  var controller = _ref.controller,
      receiver = _ref.receiver;

  var _ref2 = hooks.useAttached(receiver, receiver.attached.root),
      children = _ref2.children;

  var _controller$renderer = controller.renderer,
      renderComponent = _controller$renderer.renderComponent,
      renderText = _controller$renderer.renderText;
  return /*#__PURE__*/jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: children.map(function (child) {
      switch (child.kind) {
        case core.KIND_COMPONENT:
          return renderComponent({
            component: child,
            receiver: receiver,
            controller: controller,
            key: child.id
          });

        case core.KIND_TEXT:
          return renderText({
            text: child,
            receiver: receiver,
            key: child.id
          });

        default:
          return null;
      }
    })
  });
});

exports.RemoteRenderer = RemoteRenderer;
