// @remote-ui/rpc/build/ts/tests/utilities.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageChannel = void 0;
class MessagePortPolyfill {
    constructor() {
        this.onmessage = null;
        this.onmessageerror = null;
        this.listeners = new Set();
    }
    dispatchEvent(event) {
        if (this.onmessage) {
            this.onmessage(event);
        }
        for (const listener of this.listeners) {
            listener(event);
        }
        return true;
    }
    postMessage(message) {
        if (!this.otherPort) {
            return;
        }
        this.otherPort.dispatchEvent({ data: message });
    }
    addEventListener(type, listener) {
        if (type !== 'message') {
            return;
        }
        this.listeners.add(listener);
    }
    removeEventListener(type, listener) {
        if (type !== 'message') {
            return;
        }
        this.listeners.delete(listener);
    }
    start() { }
    close() { }
}
class MessageChannelPolyfill {
    constructor() {
        this.port1 = new MessagePortPolyfill();
        this.port2 = new MessagePortPolyfill();
        this.port1.otherPort = this.port2;
        this.port2.otherPort = this.port1;
    }
}
exports.MessageChannel = MessageChannelPolyfill;
