// @remote-ui/react/build/esm/host/RemoteRenderer.mjs
import { jsx, Fragment } from 'react/jsx-runtime';
import { memo } from 'react';
import { KIND_TEXT, KIND_COMPONENT } from '@remote-ui/core';
import { useAttached } from './hooks.mjs';

var RemoteRenderer = /*#__PURE__*/memo(function (_ref) {
  var controller = _ref.controller,
      receiver = _ref.receiver;

  var _ref2 = useAttached(receiver, receiver.attached.root),
      children = _ref2.children;

  var _controller$renderer = controller.renderer,
      renderComponent = _controller$renderer.renderComponent,
      renderText = _controller$renderer.renderText;
  return /*#__PURE__*/jsx(Fragment, {
    children: children.map(function (child) {
      switch (child.kind) {
        case KIND_COMPONENT:
          return renderComponent({
            component: child,
            receiver: receiver,
            controller: controller,
            key: child.id
          });

        case KIND_TEXT:
          return renderText({
            text: child,
            receiver: receiver,
            key: child.id
          });

        default:
          return null;
      }
    })
  });
});

export { RemoteRenderer };
