// @remote-ui/react/build/ts/host/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useWorker = exports.useAttached = exports.useRemoteReceiver = exports.ControllerContext = exports.RemoteReceiverContext = exports.createController = exports.RemoteText = exports.RemoteComponent = exports.RemoteRenderer = exports.createRemoteReceiver = void 0;
var core_1 = require("@remote-ui/core");
Object.defineProperty(exports, "createRemoteReceiver", { enumerable: true, get: function () { return core_1.createRemoteReceiver; } });
var RemoteRenderer_1 = require("./RemoteRenderer");
Object.defineProperty(exports, "RemoteRenderer", { enumerable: true, get: function () { return RemoteRenderer_1.RemoteRenderer; } });
var RemoteComponent_1 = require("./RemoteComponent");
Object.defineProperty(exports, "RemoteComponent", { enumerable: true, get: function () { return RemoteComponent_1.RemoteComponent; } });
var RemoteText_1 = require("./RemoteText");
Object.defineProperty(exports, "RemoteText", { enumerable: true, get: function () { return RemoteText_1.RemoteText; } });
var controller_1 = require("./controller");
Object.defineProperty(exports, "createController", { enumerable: true, get: function () { return controller_1.createController; } });
var context_1 = require("./context");
Object.defineProperty(exports, "RemoteReceiverContext", { enumerable: true, get: function () { return context_1.RemoteReceiverContext; } });
Object.defineProperty(exports, "ControllerContext", { enumerable: true, get: function () { return context_1.ControllerContext; } });
var hooks_1 = require("./hooks");
Object.defineProperty(exports, "useRemoteReceiver", { enumerable: true, get: function () { return hooks_1.useRemoteReceiver; } });
Object.defineProperty(exports, "useAttached", { enumerable: true, get: function () { return hooks_1.useAttached; } });
var workers_1 = require("./workers");
Object.defineProperty(exports, "useWorker", { enumerable: true, get: function () { return workers_1.useWorker; } });
