// @remote-ui/web-workers/build/ts/messenger/iframe/create-messenger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createIframeWorkerMessenger = void 0;
const constants_1 = require("./constants");
function createIframeWorkerMessenger(url, prepareIframe) {
    const { port1, port2 } = new MessageChannel();
    const iframe = document.createElement('iframe');
    iframe.setAttribute('style', 'display:none;');
    function loadHandler() {
        port1.start();
        iframe.contentWindow.postMessage({ [constants_1.IFRAME_RUN_IDENTIFIER]: url.href }, '*', [port2]);
        iframe.removeEventListener('load', loadHandler);
    }
    iframe.addEventListener('load', loadHandler);
    prepareIframe(iframe);
    document.body.appendChild(iframe);
    return {
        postMessage: (...args) => port1.postMessage(...args),
        addEventListener: (...args) => port1.addEventListener(...args),
        removeEventListener: (...args) => port1.removeEventListener(...args),
        terminate() {
            port1.close();
            iframe.remove();
        },
    };
}
exports.createIframeWorkerMessenger = createIframeWorkerMessenger;
