// @remote-ui/react/build/ts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useRemoteSubscription = exports.createRemoteReactComponent = exports.render = exports.createRemoteRoot = exports.release = exports.retain = void 0;
var rpc_1 = require("@remote-ui/rpc");
Object.defineProperty(exports, "retain", { enumerable: true, get: function () { return rpc_1.retain; } });
Object.defineProperty(exports, "release", { enumerable: true, get: function () { return rpc_1.release; } });
var core_1 = require("@remote-ui/core");
Object.defineProperty(exports, "createRemoteRoot", { enumerable: true, get: function () { return core_1.createRemoteRoot; } });
var render_1 = require("./render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return render_1.render; } });
var components_1 = require("./components");
Object.defineProperty(exports, "createRemoteReactComponent", { enumerable: true, get: function () { return components_1.createRemoteReactComponent; } });
var hooks_1 = require("./hooks");
Object.defineProperty(exports, "useRemoteSubscription", { enumerable: true, get: function () { return hooks_1.useRemoteSubscription; } });
